test/supplemental_genbank_tests/genbank_upload_full_test.py

```diff
@@ -167,3 +171,125 @@ def test_simple_shock_upload(self):
         self.assertIsNotNone(result['genome_ref'])
         self.assertTrue(int(
             result['genome_info'][10]['Number of Protein Encoding Genes']) > 0)
+
+    def _run_test_fail(self, params, error_message, mass=True):
+        with pytest.raises(Exception) as got:
+            if mass:
+                self.getImpl().genbanks_to_genomes(self.getContext(), params)
+            else:
+                self.getImpl().genbank_to_genome(self.getContext(), params)
+        assert_exception_correct(got.value, ValueError(error_message))
+
+    def _get_objects_data(self, refs):
+        results = self.getDfuClient().get_objects({'object_refs': refs})
+        return [result["info"] for result in results["data"]]
+
+    def test_genbank_to_genome_invalid_workspace(self):
+        genome_name = "GCF_000970165.1_ASM97016v1_genomic.gbff.gz"
+        params = {
+            'file': {'path': f"data/gbff/{genome_name}"},
+            'genome_name': genome_name,
+        }
+        self._run_test_fail(
+            params,
+            "Exactly one of a 'workspace_id' or a 'workspace_name' parameter must be provided",
+            False,
+        )
+
+    def test_genbanks_to_genomes(self):
```

Comment:
This should test that the data saved to the workspace is correct, including the provenance

Reply:
May I know what fields are provenance that I need to check?

Reply:
I'd just test the entire structure other than values that may change from test run to test run like timestamps, git hashes, etc

Reply:
to check the provenance, then I need to use `ws.get_objects2` instead? Also can you clarify what you mean by `test the entire structure`?

Reply:
> to check the provenance, then I need to use ws.get_objects2 instead? 

Right

> Also can you clarify what you mean by test the entire structure?

You get a provenance data structure back from the workspace. Make an equivalent data structure with the contents you expect the provenance to have and then

```
assert expected_provenance == retrieved_provenance
```

Reply:
👍

Reply:
```
            provenance = data["provenance"][0]
            retrieved_provenance = [
                {
                    'service': provenance['service'],
                    'method': provenance['method'],
                    'method_params': provenance['method_params'],
                }
            ]

            assert retrieved_provenance == expected_provenance```

This is throwing away most of the provenance data structure, and missing the parts of the structure that are the most important to capture

Reply:
It doesn't look like the object data is tested either, unless I'm missing something

Reply:
For the object info metadata test
```
assert all(info[10].get(k) == v for k, v in object_metas[idx].items())
```

Why not just
```
assert info[10] == object_metas[idx]
```

Reply:
because object_metas[idx] is a subset of info[10]

Reply:
Why is it a subset?

Reply:
Oh, another thing I missed re data checking - this is uploading assemblies so their data / prov / info needs to be checked as well

Reply:
currently only the following 5 keys are included in the object_metas. It does not have all the keys that info[10] contains. Are you suggesting we should do a full check?

            {
                "GC content": "0.41487",
                "Size": "4066551",
                "Number contigs": "1",
                "MD5": "d33802829ba0686714a5d74280527615",
                "bar": "foo",
            },

Reply:
The following is an example prov. May I know what additional fields you want to to capture? 

{'time': '2024-03-19T21:16:44+0000', 'epoch': 1710883004000, 'service': 'GenomeFileUtil', 'service_ver': 'local-dev', 'method': 'run_local_tests', 'method_params': [], 'input_ws_objects': [], 'resolved_ws_objects': [], 'intermediate_incoming': [], 'intermediate_outgoing': [], 'external_data': [], 'subactions': [{'name': 'GenomeFileUtil', 'ver': 'local-dev', 'code_url': 'https://localhost', 'commit': 'local-docker-image'}, {'name': 'DataFileUtil', 'ver': '0.2.1-release', 'code_url': 'https://github.com/kbaseapps/DataFileUtil', 'commit': '0d855560a37f8787ab3bb67e464b9c94b299764e'}, {'name': 'AssemblyUtil', 'ver': '3.1.1-release', 'code_url': 'https://github.com/kbaseapps/AssemblyUtil', 'commit': 'c40f2358ab0a17683aac8ebbdec40464299b1e47'}, {'name': 'WsLargeDataIO', 'ver': '0.0.5-beta', 'code_url': 'https://github.com/kbaseapps/WsLargeDataIO', 'commit': 'ef3871204f798844738a268005e84a0530f7aaf2'}], 'custom': {}, 'description': 'KBase SDK method run via the KBase Execution Engine'}

Reply:
Everything in there other than things that are going to change over time, which should be replaced with a placeholder value, then compare the retrieved map (with placeholders) vs. expected

Reply:
Also, what fields we should capture is data?

Reply:
Same deal, everything that isn't going to change over time, which should only be references to other objects or handles

Reply:
Do we have example/reference code how data is check? We might just want to check some fields inside. Is couple thousands lines inside at least, if not more.

Reply:
What I would do then is
* run the test but don't have it delete the workspace
* pull down the object(s) that resulted as part of the test
* Replace keys that'll change over time with place holders
* Save the objects in the test directory (pretty printed)
* In the test:
  * get the object
  * Replace keys that'll change with place holders
  * load the expected data from disk
  * `assert got == expected`

Reply:
prov is done. Will do data check in a future PR.

Reply:
I'm sorry to say it but I think the data check needs to happen in this PR. The main functionality this PR adds is uploading data to the workspace in batch, and that functionality needs to be tested before it's merged, including assembly data. 

It's actually even worse, as I forgot that the genbank and assembly files also need to be tested

Reply:
I am not sure what went wrong.

If I pull the retrieved_data down from raw logs and compare with expected data, it worked. Once moved to GHA, assertion failed.

Reply:
> pull the retrieved_data down from raw logs

I'm not quite sure what you mean by this

Reply:
Got jna problem when I tried to run test locally. Instead, I printed retrieved_data and got them from GHA raw logs.

Reply:
The GFU tests? but they're python... I'm definitely misunderstanding something

Reply:
So sdk-test fails for you?

Reply:
<img width="1728" alt="Screenshot 2024-03-27 at 6 07 51 PM" src="https://github.com/kbaseapps/GenomeFileUtil/assets/22202096/6f679752-e22f-43a4-9329-bdbfd957020a">

Reply:
You're going to need to be able to run tests locally, and the sdk is probably not going to get an update anytime soon. Probably your best bet is a VM

Reply:
does lab have any VM I can use and have docker?

Reply:
I use https://www.virtualbox.org/

Reply:
I cannot believe is free. 10 years ago I need to pay roughly $100 to put windows on my mac (dual system).

Reply:
both genome info/prov/data and assembly info/prov/data check are complete.

Reply:
:+1: